// @payments-example/lambda-utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asHttpErrorResponse = exports.parseLimit = exports.jsonResponse = exports.assertPartnerOwnership = exports.assertMerchantOwnership = exports.assertRole = exports.requireAuthContext = exports.ValidationError = exports.AuthorizationError = exports.AuthenticationError = exports.AGREEMENT_EVENT_SOURCE = exports.AGREEMENT_SETTLED_DETAIL_TYPE = exports.AGREEMENT_FUNDED_DETAIL_TYPE = exports.AGREEMENT_APPROVED_DETAIL_TYPE = exports.AGREEMENT_CREATED_DETAIL_TYPE = void 0;
// Domain event primitives shared across all Lambda packages
exports.AGREEMENT_CREATED_DETAIL_TYPE = 'AgreementCreated';
exports.AGREEMENT_APPROVED_DETAIL_TYPE = 'AgreementApproved';
exports.AGREEMENT_FUNDED_DETAIL_TYPE = 'AgreementFunded';
exports.AGREEMENT_SETTLED_DETAIL_TYPE = 'AgreementSettled';
exports.AGREEMENT_EVENT_SOURCE = 'payments-example.agreements';
class AuthenticationError extends Error {
}
exports.AuthenticationError = AuthenticationError;
class AuthorizationError extends Error {
}
exports.AuthorizationError = AuthorizationError;
class ValidationError extends Error {
}
exports.ValidationError = ValidationError;
const isSamLocal = () => process.env.AWS_SAM_LOCAL === 'true';
const getBearerToken = (event) => {
    const headers = event.headers ?? {};
    const rawHeader = headers.Authorization ?? headers.authorization;
    if (!rawHeader || typeof rawHeader !== 'string') {
        throw new AuthenticationError('Authorization bearer token is required');
    }
    const [scheme, token] = rawHeader.split(' ');
    if (scheme !== 'Bearer' || !token) {
        throw new AuthenticationError('Authorization header must use Bearer token format');
    }
    return token;
};
const parseClaimsFromBearerToken = (token) => {
    const segments = token.split('.');
    if (segments.length < 2) {
        throw new AuthenticationError('JWT bearer token format is invalid');
    }
    try {
        const payload = JSON.parse(Buffer.from(segments[1], 'base64url').toString('utf8'));
        if (!payload || typeof payload !== 'object' || Array.isArray(payload)) {
            throw new AuthenticationError('JWT bearer token payload is invalid');
        }
        return payload;
    }
    catch (error) {
        if (error instanceof AuthenticationError) {
            throw error;
        }
        throw new AuthenticationError('JWT bearer token payload is invalid');
    }
};
const getJwtClaims = (event) => {
    const authorizerClaims = event.requestContext?.authorizer?.jwt?.claims;
    if (authorizerClaims && typeof authorizerClaims === 'object') {
        return authorizerClaims;
    }
    if (!isSamLocal()) {
        throw new AuthenticationError('JWT authorizer claims are required');
    }
    return parseClaimsFromBearerToken(getBearerToken(event));
};
const buildAuthContextFromClaims = (claims) => {
    const subject = claims.sub;
    const role = claims.app_role;
    if (!subject || subject.trim() === '') {
        throw new AuthenticationError('JWT subject claim is required');
    }
    if (role !== 'merchant' && role !== 'partner' && role !== 'admin') {
        throw new AuthenticationError('JWT app_role claim is invalid');
    }
    const authContext = {
        subject,
        role,
    };
    if (role === 'merchant') {
        if (!claims.merchant_id || claims.merchant_id.trim() === '') {
            throw new AuthenticationError('JWT merchant_id claim is required for merchants');
        }
        authContext.merchantId = claims.merchant_id;
    }
    if (role === 'partner') {
        if (!claims.partner_id || claims.partner_id.trim() === '') {
            throw new AuthenticationError('JWT partner_id claim is required for partners');
        }
        authContext.partnerId = claims.partner_id;
    }
    return authContext;
};
const requireAuthContext = (event) => buildAuthContextFromClaims(getJwtClaims(event));
exports.requireAuthContext = requireAuthContext;
const assertRole = (authContext, expectedRole, message) => {
    if (authContext.role !== expectedRole) {
        throw new AuthorizationError(message);
    }
};
exports.assertRole = assertRole;
const assertMerchantOwnership = (authContext, merchantId, message) => {
    if (authContext.merchantId !== merchantId) {
        throw new AuthorizationError(message);
    }
};
exports.assertMerchantOwnership = assertMerchantOwnership;
const assertPartnerOwnership = (authContext, partnerId, message) => {
    if (authContext.partnerId !== partnerId) {
        throw new AuthorizationError(message);
    }
};
exports.assertPartnerOwnership = assertPartnerOwnership;
const jsonResponse = (statusCode, body) => ({
    statusCode,
    body: JSON.stringify(body),
});
exports.jsonResponse = jsonResponse;
const parseLimit = (event, options) => {
    const defaultLimit = options?.default ?? 50;
    const maxLimit = options?.max ?? 200;
    const rawLimit = event.queryStringParameters?.limit;
    if (!rawLimit) {
        return defaultLimit;
    }
    const parsed = Number(rawLimit);
    if (!Number.isInteger(parsed) || parsed <= 0 || parsed > maxLimit) {
        throw new ValidationError(`limit must be an integer between 1 and ${maxLimit}`);
    }
    return parsed;
};
exports.parseLimit = parseLimit;
const asHttpErrorResponse = (error) => {
    if (error instanceof ValidationError) {
        return (0, exports.jsonResponse)(400, { message: error.message });
    }
    if (error instanceof AuthenticationError) {
        return (0, exports.jsonResponse)(401, { message: error.message });
    }
    if (error instanceof AuthorizationError) {
        return (0, exports.jsonResponse)(403, { message: error.message });
    }
    return null;
};
exports.asHttpErrorResponse = asHttpErrorResponse;
